catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/security/SystemTemplateAccessPlugin.java

```diff
@@ -0,0 +1,154 @@
+import ddf.catalog.operation.ResourceRequest;
+import ddf.catalog.operation.ResourceResponse;
+import ddf.catalog.operation.UpdateRequest;
+import ddf.catalog.plugin.AccessPlugin;
+import ddf.catalog.plugin.StopProcessingException;
+import java.io.Serializable;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import org.codice.ddf.catalog.ui.util.EndpointUtil;
+import org.codice.ddf.security.common.Security;
+
+public class SystemTemplateAccessPlugin implements AccessPlugin {
+  private final EndpointUtil util;
+
+  private final Supplier<Boolean> subjectHasAdmin;
+
+  public SystemTemplateAccessPlugin(EndpointUtil util) {
+    this(util, () -> Security.getInstance().javaSubjectHasAdminRole());
+  }
+
+  public SystemTemplateAccessPlugin(EndpointUtil util, Supplier<Boolean> subjectHasAdmin) {
+    this.util = util;
+    this.subjectHasAdmin = subjectHasAdmin;
+  }
+
+  /**
+   * Only an admin can create system templates. This is an extra layer of protection in case the
+   * policy for shareable metacards changes or malicious users try to use a different endpoint.
+   *
+   * @param input the {@link CreateRequest} to process.
+   * @return the same {@link CreateRequest} if none of the metacards to create are considered system
+   *     templates, or if the current subject is an admin.
+   * @throws StopProcessingException if any of the metacards to create are system templates and the
+   *     current subject is not an admin.
+   */
+  @Override
+  public CreateRequest processPreCreate(CreateRequest input) throws StopProcessingException {
+    if (subjectHasAdmin.get()) {
+      return input;
+    }
+    if (input
+        .getMetacards()
+        .stream()
+        .map(Metacard::getTags)
+        .flatMap(Set::stream)
+        .anyMatch(SYSTEM_TEMPLATE::equals)) {
+      throw new StopProcessingException("System templates can only be created by an admin");
+    }
+    return input;
+  }
+
+  /**
+   * No one is allowed to update system templates.
+   *
+   * @param input the {@link UpdateRequest} to process.
+   * @param existingMetacards the Map of {@link Metacard}s that currently exist.
+   * @return the same {@link UpdateRequest} if no system templates are targets for the update.
+   * @throws StopProcessingException if any of the metacards targeted for the update are system
+   *     templates.
+   */
+  @Override
+  public UpdateRequest processPreUpdate(
+      UpdateRequest input, Map<String, Metacard> existingMetacards) throws StopProcessingException {
+    Set<String> systemTemplateIds =
+        util.getMetacardsByFilter(SYSTEM_TEMPLATE)
+            .values()
+            .stream()
+            .map(Result::getMetacard)
+            .map(Metacard::getId)
+            .collect(Collectors.toSet());
+    if (existingMetacards.keySet().stream().anyMatch(systemTemplateIds::contains)) {
+      throw new StopProcessingException("System templates cannot be updated");
```

Comment:
❓ Can we just check if any of the metacards have the `SYSTEM_TEMPLATE` tag? (instead of querying for all ID's then comparing them?)